src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveSummaryService.kt

```diff
@@ -15,70 +18,63 @@ class IncentiveSummaryService(
   private val prisonApiService: PrisonApiService
 ) {
 
-  fun getIncentivesSummaryByLocation(prisonId: String, locationId: String, sortBy: SortColumn = SortColumn.NAME, sortDirection: Sort.Direction = Sort.Direction.ASC): Mono<BehaviourSummary> {
-
-    val result = prisonApiService.findPrisonersAtLocation(prisonId, locationId)
-      .collectList()
-      .map {
-
-        val bookingIds = it.map { p -> p.bookingId }.toList()
-        val offenderNos = it.map { p -> p.offenderNo }.toList()
-
-        if (bookingIds.isEmpty()) throw NoPrisonersAtLocationException(prisonId, locationId)
-
-        Mono.zip(
-          Mono.just(it),
-          getIEPDetails(bookingIds),
-          getCaseNoteUsage("POS", "IEP_ENC", offenderNos),
-          getCaseNoteUsage("NEG", "IEP_WARN", offenderNos),
-          getProvenAdjudications(bookingIds),
-          getIepLevelsByDescription(prisonId)
-        ).map { tuples ->
-          getPrisonersByLevel(tuples.t1, tuples.t2)
-            .map { prisonerIepLevelMap ->
-              val iepLevel = lookupIepLevel(prisonerIepLevelMap, tuples.t6)
-              IncentiveLevelSummary(
-                level = iepLevel.iepLevel,
-                levelDescription = iepLevel.iepDescription,
-                prisonerBehaviours = prisonerIepLevelMap.value.map { p ->
-                  PrisonerIncentiveSummary(
-                    firstName = WordUtils.capitalizeFully(p.firstName),
-                    lastName = WordUtils.capitalizeFully(p.lastName),
-                    prisonerNumber = p.offenderNo,
-                    bookingId = p.bookingId,
-                    imageId = p.facialImageId,
-                    daysOnLevel = tuples.t2[p.bookingId]?.daysOnLevel ?: 0,
-                    daysSinceLastReview = tuples.t2[p.bookingId]?.daysSinceReview ?: 0,
-                    positiveBehaviours = tuples.t3[p.offenderNo]?.totalCaseNotes ?: 0,
-                    incentiveEncouragements = tuples.t3[p.offenderNo]?.numSubTypeCount ?: 0,
-                    negativeBehaviours = tuples.t4[p.offenderNo]?.totalCaseNotes ?: 0,
-                    incentiveWarnings = tuples.t4[p.offenderNo]?.numSubTypeCount ?: 0,
-                    provenAdjudications = tuples.t5[p.bookingId]?.provenAdjudicationCount ?: 0,
-                  )
-                }.sortedWith(sortBy.applySorting(sortDirection))
+  suspend fun getIncentivesSummaryByLocation(
+    prisonId: String,
+    locationId: String,
+    sortBy: SortColumn = SortColumn.NAME,
+    sortDirection: Sort.Direction = Sort.Direction.ASC
+  ): BehaviourSummary {
+
+    val prisoners = prisonApiService.findPrisonersAtLocation(prisonId, locationId).toList()
+    val bookingIds = prisoners.map { p -> p.bookingId }.toList()
+    val offenderNos = prisoners.map { p -> p.offenderNo }.toList()
+
+    if (bookingIds.isEmpty()) throw NoPrisonersAtLocationException(prisonId, locationId)
```

Comment:
would it be same to check whether `prisoners` was `empty` instead, before the two `bookingIds` and `offenderNos` creation

Reply:
Updated